### ISSUE TITLE
How to use float BatchNorm in ZeRO optimizer?

### ISSUE THREAD

**Issue by yanring** [microsoft/DeepSpeed#520]
I'm trying to use mixed precision in ZeRO, mainly because I need float BatchNorm layers, so is there an example or documentation for this?

**Comment by samyam** [comment]
@Yanring ZeRO currently only supports fp16 parameters, and we are thinking about extending it to support a mixture of both fp16 and fp32 parameters. But we do not have an ETA yet.

However, you can still do float BatchNorm  without requiring mixed precision in ZeRO. BatchNorm consists of two parts: i) the actual normalization part (shown in blue in the figure below) and ii) the affine part which requires parameters (shown in red below). The first part needs to be done in float, since there is a reduction operation involved, but the second part can be done in fp16, without any impact to accuracy or stability. So you can do your batch norm in fp32, then convert the activation to fp16 before doing the affine part. This way you can still use ZeRO while using float BatchNorm computation.

Please let us know if you have more questions on this.
 
![image](https://user-images.githubusercontent.com/3409344/98723777-adc5be80-2347-11eb-9d4e-d53316e1100c.png)

**Comment by yanring** [comment]
> @yanring ZeRO currently only supports fp16 parameters, and we are thinking about extending it to support a mixture of both fp16 and fp32 parameters. But we do not have an ETA yet.
> 
> However, you can still do float BatchNorm without requiring mixed precision in ZeRO. BatchNorm consists of two parts: i) the actual normalization part (shown in blue in the figure below) and ii) the affine part which requires parameters (shown in red below). The first part needs to be done in float, since there is a reduction operation involved, but the second part can be done in fp16, without any impact to accuracy or stability. So you can do your batch norm in fp32, then convert the activation to fp16 before doing the affine part. This way you can still use ZeRO while using float BatchNorm computation.
> 
> Please let us know if you have more questions on this.
> 
> ![image](https://user-images.githubusercontent.com/3409344/98723777-adc5be80-2347-11eb-9d4e-d53316e1100c.png)

Thanks!
